iris-mpc-cpu/src/execution/hawk_main.rs

```diff
@@ -482,37 +482,98 @@ impl HawkActor {
                 })
             });
         tracing::info!(
-            "Keeping {} distances for eye {side} out of {} search results.",
+            "Keeping {} distances for eye {side} out of {} search results. Cache size: {}/{}",
             distances.clone().count(),
-            search_results.len()
+            search_results.len(),
+            self.distances_cache[side].len(),
+            self.args.match_distances_buffer_size,
         );
         self.distances_cache[side].extend(distances);
     }
 
-    async fn compute_buckets(&self, session: &mut Session, side: usize) -> Result<VecBuckets> {
+    async fn compute_buckets(
+        &self,
+        sessions: &BothEyes<Vec<HawkSessionRef>>,
+        side: usize,
+    ) -> Result<VecBuckets> {
         let translated_thresholds = Self::calculate_threshold_a(self.args.n_buckets);
-        let bucket_result_shares = compare_threshold_buckets(
-            session,
-            translated_thresholds.as_slice(),
-            self.distances_cache[side].as_slice(),
-        )
-        .await?;
+        let n_sessions = sessions[side].len();
+        let distances = self.distances_cache[side].as_slice();
+        let chunk_size = distances.len() / n_sessions;
+        let remainder = distances.len() % n_sessions;
+
+        let mut start = 0;
+        let mut slices_data = Vec::with_capacity(n_sessions);
+        for i in 0..n_sessions {
+            let size = chunk_size + (i < remainder) as usize;
+            let end = start + size;
+            slices_data.push(distances[start..end].to_vec());
+            start = end;
+        }
+
+        let (tx, mut rx) = mpsc::unbounded_channel::<(usize, Vec<Share<u32>>)>();
+        let tasks = sessions[side]
+            .iter()
+            .cloned()
+            .zip(slices_data.into_iter())
+            .enumerate()
+            .map(|(i_session, (session_ref, slice))| {
+                let thresholds = translated_thresholds.clone();
+                let tx = tx.clone();
+                async move {
+                    let mut session = session_ref.write().await;
+                    let shares = compare_threshold_buckets(
```

Comment:
If `distances.len() < n_sessions` then sometimes `slice.len() = 0` which is not accepted by `compare_threshold_buckets`.

Reply:
Thanks, addressed 👍